### ISSUE TITLE
Passing mutable slice from JS to Rust

### ISSUE THREAD

**Issue by dignifiedquire** [alexcrichton/wasm-bindgen#53]
I am trying to reduce allocation overhead and avoid the use of `Vec`, which leads me to a C-style interface like this

```rust
#[wasm-bindgen]
pub fn expensive_work(input: &[u8], output: &mut [u8]) {
  // do expensive work
}
```

now currently this throws when trying to compile with a missmatched types error. 
Would it be possible to support something like this?

**Comment by sendilkumarn** [comment]
Interesting use case 🤔 (extremely interesting stuff) 

Does this also means that it should allow streaming something from JS to Rust

**Comment by dignifiedquire** [comment]
(for anyone interested, the actual expensive work I am doing currently is aes)

**Comment by MaxGraey** [comment]
Yes, also thinking about this approach. Sometimes cost of interop much bigger than actual work, especially for buffers.

**Comment by RReverser** [comment]
I don't think this would be possible without heavy modification of WASM (which will likely slow it down and end up copying all the modified chunks anyway) because WASM doesn't have access to memory of buffers created in JS.

Returning vector or a reference to pre-allocated buffer on Rust side might be more feasible.

**Comment by dignifiedquire** [comment]
@RReverser I don't mean to reuse the javascript land buffer. 
What I have done by hand before is simply allocating the buffer in the webassembly memory, passing the pointer, length pair to the function, and then afterwards reading it out from js, e.g. similar to how complex return types are implemented. 

So the javascript wrapping side would do this work and probably expose it as a return value, but there is a difference for me, as I can get away with fixed sizes from js, but treat all of them as slices  on the rust side.

**Comment by RReverser** [comment]
Sorry, I don't fully understand, can you share the piece of JS you're currently using and that you expect to get rid of?

**Comment by MaxGraey** [comment]
I think about this as reuse memory as much as possible. For example:
- request allocation in AS linear memory via js side by rust's allocator (need export malloc/alloc and free from rust)
- get pointer after that (still in js side)
- pass this pointer to rust side
- rust side unwrap to mem slice (unsafe)
- do write or read (rust side)
- now we can read from the same chunk if needed or manually free this memory (js side)

Is it possible?

**Comment by dignifiedquire** [comment]
Roughly what @MaxGraey described. I did write this once by hand I can see if I can dig up the code later.

**Comment by RReverser** [comment]
So if I understand correctly, the goal is to reuse memory to return results. What I don't understand so far, is why is it needed to manage that memory (alloc, dealloc) from JS and pass pointers to Rust, when one can just do this all in Rust either via regular `static` for buffers with known size, or with `lazy_static!` for buffers without one, and just always return `&'static [u8]` to JS that points to the same backing buffer.

Seems like that should be easier to work with, as it doesn't involve allocation hacks on JS side?

**Comment by MaxGraey** [comment]
> So if I understand correctly, the goal is to reuse memory to return results

Exactly

> What I don't understand so far, is why is it needed to manage that memory (alloc, dealloc) from JS and pass pointers to Rust

I think copy memory from TypedArray to AS memory (actually BufferView) much faster than on WASM (without bulk memory operation which not implemented yes). If I wrong may this steps is unnecessary and we can do in on rust side.

**Comment by dignifiedquire** [comment]
It's currently not possible to return `&'static [u8]` using bindgen as far as I can tell, so at a minimum that would need to be supported. 

On another note, one of the reasons I want to do this is to avoid having to include vec in my implementation to keep the final file size as small as possible. The trick here is that the dynamic allocation part at runtime would happen on the js side where I can easily do it, without having to do it on the rust side.
If there are ways to achieve this differently, I am all ears.

**Comment by alexcrichton** [comment]
Thanks for the report @dignifiedquire and the info @MaxGraey! I wonder though if this use case (not copying data to/from buffers) would be best served with something like:

```rust
#[wasm_bindgen]
pub struct MyResult {
   data: Vec<u8>,
}

#[wasm_bindgen]
impl MyResult {
    pub fn add(&mut self, data: &[u8]) { /* ... */ }

    pub fn ptr(&self) -> *const u8 { /* ... */ }
    pub fn len(&self) -> usize { /* ... */ }
}
```

That way all the data would stay owned by Rust and you wouldn't have to copy data in/out?

Right now a binding for `&mut [u8]` would probably be something that takes a `Uint8Array` in JS, copies it to Rust, calls Rust, and then copies it back out upon returning. I'd imagine though that all that copying isn't quite what you want?

**Comment by MaxGraey** [comment]
Yes. We can do it better I think, without coping result. Something like this:

```js
const { MyResult } = module.exports;
MyResult.add(...);
const resultPtr  = MyResult.ptr();
const resultLen = MyResult.len();
const result = new Uint8Array(module.HEAP8.buffer, resultPtr, resultLen);
... do some stuff with result
MyResult.free(resultPtr);
```

Or with clone:

```js
...
wrappedCall(...) {
  ... the same as above ...
  const result = new Uint8Array(module.HEAP8.buffer, resultPtr, resultLen);
  const cloned = new Uint8Array(result);
  MyResult.free(resultPtr);
  return cloned;
}

const result = module.wrappedCall(..)
... do some stuff with result
```

Last one we can wrap all free stuff into interop call, but we should do one copy.
The basic idea. Do copy on JS side (probably this faster then on WASM without bulk operations). And do it in more flexible way - with or without copying result.

**Comment by MaxGraey** [comment]
> It's currently not possible to return &'static [u8] using bindgen as far as I can tell, so at a minimum that would need to be supported.

I'm pretty new in rust, but I think &'static [u8] ref to static data segment which is not perfect area for large and frequently change data.

**Comment by RReverser** [comment]
@MaxGraey No, `&'static` is a reference to any global data, including variables. So, given WASM can be used only from single thread in JS, you can put some data in such global array, and return a reference or pointer to JS which can do whatever it wants with data before calling into WASM again.

**Comment by alexcrichton** [comment]
@MaxGraey oh I've had pretty bad luck using long-lived typed array views into wasm memory, are you sure that's supported? Whenever I've done that I've ended up needing to copy it into JS b/c Rust can stomp all over the freed memory.

**Comment by RReverser** [comment]
@alexcrichton yeah that works fine if you are processing data immediately (or at least before making next call into WASM).
